common/migration_mechanism.py

```diff
@@ -0,0 +1,192 @@
+    return glob.glob(os.path.join(data_root, 'players', '%s_*.json' % player_name), recursive=True)
+
+
+def _load_datastores(data_root: str, player_name: str) -> Dict[str, Dict]:
+    # Find datastores for the player
+    player_datastores = _get_datastores_for_player(data_root, player_name)
+    result = OrderedDict()
+    # Load each of them under the top-level dict
+    for ds in player_datastores:
+        datastore_name: str = os.path.splitext(os.path.basename(ds))[0].split('_')[-1]
+        with open(ds, 'rt') as f:
+            result[datastore_name] = json.load(f)
+    return result
+
+
+def _save_datastores(data_root: str, player_name: str, player_data: Dict[str, Dict]) -> None:
+    for ds_name, ds_data in player_data.items():
+        file_path = os.path.join(data_root, 'players', '%s_%s.json' % (player_name, ds_name))
+        with open(file_path, 'wt') as f:
+            json.dump(ds_data, f, indent=4, sort_keys=True)
+
+
+# Migrates all files of all schemas
+def run_migrations(data_root_path: str) -> None:
+    """
+    Run data migrations on all necessary files
+
+    This does no error checking; if any migration fails a ValueError will be raised at that point
+
+    :param data_root_path: the root path of data files to migrate
+    :return: None
+    """
+    existing_version = _load_schema_version(data_root_path)
+
+    # Determine the highest available migration
+    upgraded_version = existing_version
+    while upgraded_version + 1 in _registered_migrations:
+        upgraded_version += 1
+
+    # Exit early if no migration is needed
+    if upgraded_version == existing_version:
+        return
+
+    # Back up all datastores
+    _perform_backups(data_root_path, existing_version)
+
+    # Perform each migration in turn
+    for i in range(existing_version + 1, upgraded_version + 1):
+        _registered_migrations[i]()
+
+    # Write the new schema version
+    _save_schema_version(data_root_path, upgraded_version)
+
+
+# Migration decorator
+def taserver_migration(schema_version: int):
+    """
+    Decorator denoting a TAServer schema migration
+    Should decorate a function taking no arguments and returning nothing. The migration function is essentially a script
+    performing whatever migration required
+
+    Migrations should raise a ValueError if the data they attempt to upgrade is invalid
+
+    :param schema_version: the version of the schema this migration upgrades to
+                           there should only be one migration for a given version of a given schema
+                           and there must be a contiguous range of migrations available, starting at version 1
+    """
+    def decorator(func):
+        @wraps(func)
+        def wrapped_func():
+            return func()
+        # Register the migration
+        _registered_migrations[schema_version] = func
+        return wrapped_func
+    return decorator
+
+
+def upgrades_all_players(data_root_path: str):
```

Comment:
What's the reason for passing the data root through the decorator?

Reply:
Removed; not needed due to the data root now being passed into migration funcs (as per other comment)